concrete/blocks/core_area_layout/controller.php

```diff
@@ -23,6 +23,77 @@ class Controller extends BlockController
     protected $btSupportsInlineEdit = true;
     protected $btTable = 'btCoreAreaLayout';
     protected $btIsInternal = true;
+    protected $btCacheSettingsInitialized = false;
+
+    public function cacheBlockOutput()
+    {
+        $this->setupCacheSettings();
+
+        return $this->btCacheBlockOutput;
+    }
+
+    public function cacheBlockOutputOnPost()
+    {
+        $this->setupCacheSettings();
+
+        return $this->btCacheBlockOutputOnPost;
+    }
+
+    public function getBlockTypeCacheOutputLifetime()
+    {
+        $this->setupCacheSettings();
+
+        return $this->btCacheBlockOutputLifetime;
+    }
+
+    protected function setupCacheSettings()
+    {
+        if ($this->btCacheSettingsInitialized || Page::getCurrentPage()->isEditMode()) {
+            return;
+        }
+
+        $this->btCacheSettingsInitialized = true;
+
+        $btCacheBlockOutput = true;
+        $btCacheBlockOutputOnPost = true;
+        $btCacheBlockOutputLifetime = 0;
+
+        $c = $this->getCollectionObject();
+
+        $blocks = [];
+        $layout = $this->getAreaLayoutObject();
+        $layout->setAreaObject($this->getAreaObject());
+        if ($layout) {
+            foreach($layout->getAreaLayoutColumns() as $column) {
+                $area = $column->getSubAreaObject();
+                if ($area) {
+                    foreach($area->getAreaBlocksArray($c) as $block) {
```

Comment:
@aembler another addition I needed to make this work in a predictable manner.
```php
// Check permissions because visibility becomes unpredictable and also cache settings may otherwise reveal information about contents.
$ap = new \Permissions($area);
if (!$ap->canViewArea()) {
    continue;
}
```

Reply:
Hmm I can see what you mean but this isn't the right way to do it. What happens if the first time it's rendered you're running this as the super user? You're going to be able to view the area and it'll render into cache even if guest users shouldn't see it. This is one of the problems of trying to make this caching work.

Reply:
We're updating our sites with 8.4 and are wondering if you solved block visibility (user permissions) in combination with area layout caching?